assembly/indexer/ProtoruneRuneId.ts

```diff
@@ -0,0 +1,57 @@
+import { u128 } from "as-bignum/assembly";
+import { RuneId } from "metashrew-runes/assembly/indexer/RuneId";
+import { readULEB128ToU128 } from "metashrew-runes/assembly/leb128.ts";
+import { toLEB128, concatByteArrayTruncateZeros, logArrayBuffer } from "../utils";
+import { Box } from "metashrew-as/assembly/utils/box";
+import { console } from "metashrew-as/assembly/utils/logging";
+
+export class ProtoruneRuneId extends RuneId {
+  decode(): Array<u128> {
+    const fieldData = new Array<u128>(2);
+    fieldData[0] = this.block;
+    fieldData[1] = this.tx;
+    const input = Box.from(concatByteArrayTruncateZeros(fieldData));
+    const result: Array<u128> = new Array<u128>();
+    while (input.len > 0) {
+      const current = u128.from(0);
+      let size = readULEB128ToU128(input, current);
+      if (size === usize.MAX_VALUE) return changetype<Array<u128>>(0); //can choose to continue or return
+      input.shrinkFront(size);
+      result.push(current);
+    }
+    return result;
+  }
+  // ary should have the string identifier for the protorune first, then the data that should be encoded
+  // Ex: ["OYL", RuneBlock1, RuneTx1, RuneBlock2, RuneTx2]. Guarantees that the padded bits can be ignored and the first data byte is nonzero (must be a string)
+  // Important note: collisions with existing RuneIds are not explicitly handled. Authors of subprotocols should ensure that the ProtoruneRuneId scheme does not conflict with existing rune ids.
```

Comment:
hm, if we can't guarantee that protorunrRuneIds are unique, then there could be attacks on subprotocols.

For example, if oylswap mints a protorune with ProtoruneRuneId as an LP token, someone can just copy the params called here to mint their own protorune that will be indistinguishable from the real one

Reply:
As discussed in chat, this should be implemented in the subprotocol